### DISCUSSION TITLE
Is there any way to add values to an array column with UPDATE?

### DISCUSSION THREAD

**Discussion by ru88s** [supabase/supabase#1570] [Questions] [answered]
I want to add a value to an array column using update.
[31] → [31,32]
![スクリーンショット 2021-05-15 21 56 56](https://user-images.githubusercontent.com/42270511/118362311-02cde000-b5ca-11eb-8d63-8a059f1e7c42.png)

**Comment by easy-street** [comment, marked as answer]
two options currently:

1) [update](https://supabase.io/docs/reference/javascript/update) with the complete new array
```
await supabase
  .from("foos")
  .update({ refer_comment: [31, 32] })
  .eq("id", 1);
```

2) [rpc](https://supabase.io/docs/reference/javascript/rpc) with a function that accepts the value to append to the array
```
await supabase.rpc("append_comment_id", {
  foo_id: 1,
  comment_id: 32,
});
```
from the screenshot, it looks like the column is defined as a JSON type, not as an Array type (`int[]`), correct? example `append_comment_id` function assuming `refer_comment` is defined as type `jsonb`:
```
create function append_comment_id(foo_id int, comment_id int) returns foos as $$
  update foos
  set refer_comment = coalesce(refer_comment, '[]'::jsonb) || to_jsonb(comment_id)
  where id = foo_id
  returning *;
$$ language sql;
```

**Reply by ru88s** [reply]
Thank you very much.
And the json type was a mistake.
I fixed it to array type. Sorry about that.
What is the method for the array type?
![スクリーンショット 2021-05-17 9 54 06](https://user-images.githubusercontent.com/42270511/118419133-255d1780-b6f6-11eb-8486-6a8298ad362c.png)

**Reply by easy-street** [reply]
example `append_comment_id` function when `refer_comment` is defined as type `int[]`:

```
create function append_comment_id(foo_id int, comment_id int) returns foos as $$
  update foos
  set refer_comment = refer_comment || comment_id
  where id = foo_id
  returning *;
$$ language sql;
```

**Reply by ru88s** [reply]
Thank you very much!
I was able to register with Funciton.

How can I represent `in` in a function?
I would like to achieve the following in a function.

```
const { data: commentData } = await supabase
      .from("comments")
      .update({ refer_comment: next_res_Id })
      .match({ talk_id: talk_id })
      .in("res_id", anker_array);
```

**Reply by easy-street** [reply]
seems like you're on the right track. working example:

```
const { data: commentData } = await supabase
  .from("comments")
  .update({ refer_comment: [31, 32] })
  .match({ talk_id: 1 })
  .in("res_id", [1, 2]);
```

**Reply by ru88s** [reply]
I'm sorry.
I would like to know how to do the same thing in SQL.

Thank you for helping me again and again.

The following does not work.

```
create function append_comment_id(next_res_id int, talk_id int, ankers array) 
returns void as $$
  update comments
  set refer_comment = refer_comment || next_res_id
  where id = talk_id
  res_id IN ankers
$$ language sql volatile;
```
![スクリーンショット 2021-05-17 17 18 34](https://user-images.githubusercontent.com/42270511/118454993-eb126b00-b733-11eb-998a-00a5fdd7982e.png)

**Reply by easy-street** [reply]
looks like just a couple adjustments are needed:
```
create function append_comment_id(next_res_id int, talk_id int, ankers int[])
returns comments as $$
  update comments
  set refer_comment = refer_comment || next_res_id
  where id = talk_id
  and res_id = any(ankers)
  returning *;
$$ language sql;
```

**Reply by ru88s** [reply]
It returned null.
Am I doing something wrong?

```
const { data, error } = await supabase.rpc("append_comment_id", {
      ankers: new_anker_nos,
      talk_id: talk.id,
      next_res_id: nextResId,
    });
```

```
create function append_comment_id(next_res_id int, talk_id int, ankers int[]) 
returns comments as $$
  update comments
  set refer_comment = refer_comment || next_res_id
  where id = talk_id
  and res_id = any(ankers)
  returning *;
$$ language sql;
```

**Reply by easy-street** [reply]
looks correct all around. the behavior you're describing indicates that there is not a row satisfying the predicates. only the row with both of the following will be updated:
- column `id` of type `int` with value equal to the value of `talk.id`
- column `res_id` of type `int` with value equal to any of the integers in `new_anker_nos`

you might try testing with values instead of variables. for example, this works as expected, since a row exists where `id` is `1` and `res_id` is `1`:
```
const { data, error } = await supabase.rpc('append_comment_id', {
  ankers: [1,2],
  talk_id: 1,
  next_res_id: 32,
});
```

**Reply by ru88s** [reply]
I'm sorry.
I don't know why it's not working.

<img width="565" alt="スクリーンショット 2021-05-18 10 06 27" src="https://user-images.githubusercontent.com/42270511/118575272-11321c80-b7c1-11eb-8afb-8e30f8c23938.png">

![スクリーンショット 2021-05-18 10 06 42](https://user-images.githubusercontent.com/42270511/118575277-12fbe000-b7c1-11eb-9307-d8a9e07d3622.png)

![スクリーンショット 2021-05-18 10 06 49](https://user-images.githubusercontent.com/42270511/118575280-12fbe000-b7c1-11eb-8048-5ddd843944a3.png)
![スクリーンショット 2021-05-18 10 07 04](https://user-images.githubusercontent.com/42270511/118575283-13947680-b7c1-11eb-86af-312fb1b66db8.png)

**Reply by easy-street** [reply]
this behavior is expected when **there is no row** in the `comments` table with these values:
`id`: `1`
`res_id`: `1` or `2`

**Reply by ru88s** [reply]
The value is present, but fails.

![スクリーンショット 2021-05-18 10 22 53](https://user-images.githubusercontent.com/42270511/118576229-14c6a300-b7c3-11eb-85c5-e6b197ae2008.png)
<img width="571" alt="スクリーンショット 2021-05-18 10 23 10" src="https://user-images.githubusercontent.com/42270511/118576233-155f3980-b7c3-11eb-9fbf-eca9cbd5291e.png">

**Reply by easy-street** [reply]
so, in [this comment](https://github.com/supabase/supabase/discussions/1570#discussioncomment-746910), you included this predicate:
```
where id = talk_id
```

but in the screenshot here, you're showing the `talk_id` column, not the `id` column.

if you want the predicate to reference the `talk_id` column rather than the `id` column, you'll need to change the predicate:
```
where talk_id = append_comment_id.talk_id
```
(qualifying the `talk_id` argument with the name of the function to disambiguate vs. the `talk_id` _column_. alternatively, you could change the name of the argument.)

so the complete function would be:
```
create function append_comment_id(next_res_id int, talk_id int, ankers int[]) returns comments as $$
  update comments
  set refer_comment = refer_comment || next_res_id
  where talk_id = append_comment_id.talk_id
  and res_id = any(ankers)
  returning *;
$$ language sql;
```

**Reply by ru88s** [reply]
It was my mistake. I'm sorry!
It's done!
Thanks a million!

**Comment by emjayess** [comment]
I was attempting to use one or another of the `jsonb_*` functions supposedly supported by pgsql, and even suggested by "supabase AI" in the sql editor window, and they just failed as unsupported or unknown functions.

**Reply by Uyiosa2607** [reply]
The same issue I am having

**Reply by GaryAustin1** [reply]
You both are commenting on a 2 year old thread.
You should state what your problem is in a new topic here or Discord.
You cannot do partial updates of arrays or json thru the REST API.  Those have to be with an rpc call to a Postgres function.
